### PR TITLE
Make third argument of `DeclareRepresentation` and `NewRepresentation` optional, and document that it and the fourth argument are (and always were) unused

### PR DESCRIPTION
**Pull request by fingolfin**
See the commit messages for details.

See also issues #4409 and #1042

### CODE REVIEW

File: lib/type.g [hunk 1]
```diff
@@ -190,26 +190,13 @@ end );
...
-##  subrepresentation of it, and <A>slots</A> tells what positions of the
-##  objects in the representation <A>rep</A> may be bound.
-##  In the latter case,
-##  <A>rep</A> must be <Ref Filt="IsComponentObjectRep"/> or a
-##  subrepresentation of, and <A>slots</A> lists the admissible names of
-##  components that objects in the representation <A>rep</A> may have.
-##  The admissible positions resp. component names of <A>super</A> need not
-##  be listed in <A>slots</A>.
+##  The optional third and fourth arguments <A>slots</A> and <A>req</A>
+##  are unused (and never were) only provided for backwards compatibility.
```
Comment (ThomasBreuer) [inline lib/type.g:194]: perhaps "are (and always were) unused and only provided for backwards compatibility"

**Review by ThomasBreuer — COMMENTED** [review]
I agree with the changes, and have made just one comment concerning a formulation.
Concerning the newly inserted `# TODO` comments:
I think they might be confusing for readers in the future.
The filters in question have not been used before, in particular they were not intended to be automatically implied by the newly created filters.

File: lib/type.g (hunk 1 above)
Comment (fingolfin) [reply lib/type.g]: Indeed, that's much better. Fixed now

**Comment by fingolfin** [comment]
The reason I added those TODO comments is because without them, it feels as if some information is lost... and yeah, I know these were not implied, but I still wonder why? I guess this is the same underlying issue as in PR #4499: Why doesn't `IsBlockMatrixRep` not imply `IsMatrix`? It is genuinely unclear to me. Is it really just "purity", or a deeper thing? The argument that one might want to use `IsBlockMatrixRep` for something that is not a matrix doesn't strike me as convincing -- I mean, it is *in the name* that this is for matrices. One argument of course is that nowadays, perhaps it should be `IsMatrixOrMatrixObj`. But other than that? 

On the other hand, we have had plenty of ranking issues (esp. with "hidden implications" disabled) because people used (and use) representations in method argument filters, unaware that these filters lack many seemingly "obvious" implications.


Anyway, I don't think this particular instance is worth arguing about a lot, so I just dropped these comments now *shrug*

**Review by ThomasBreuer — APPROVED** [review]

**Pull request merged**